File: docs/examples/go/src/data_management_attachments.go
package main

import (
	"context"

	reduct "github.com/reductstore/reduct-go"
)

func main() {
	// Create a client and use the base URL and API token
	client := reduct.NewClient("http://localhost:8383", reduct.ClientOptions{
		APIToken: "my-token",
	})

	// Get or create a bucket with the name "my-bucket"
	bucket, err := client.CreateOrGetBucket(context.Background(), "my-bucket", nil)
	if err != nil {
		panic(err)
	}

	// Write attachments for an entry
	err = bucket.WriteAttachments(context.Background(), "camera/front", map[string]any{
		"schema":      map[string]any{"type": "object", "version": "1.0"},
		"calibration": map[string]any{"fx": 1260.1, "fy": 1261.2},
	})
	if err != nil {
		panic(err)
	}

	// Read attachments
	attachments, err := bucket.ReadAttachments(context.Background(), "camera/front")
	if err != nil {
		panic(err)
	}
	schema, ok := attachments["schema"].(map[string]any)
	if !ok || schema["type"] != "object" {
		panic("schema attachment was not read correctly")
	}

	// Remove one attachment by key
	err = bucket.RemoveAttachments(context.Background(), "camera/front", []string{"calibration"})
	if err != nil {
		panic(err)
	}

	attachments, err = bucket.ReadAttachments(context.Background(), "camera/front")
	if err != nil {
		panic(err)
	}
	if _, exists := attachments["calibration"]; exists {
		panic("calibration attachment must be removed")
	}

	// Remove all attachments
	err = bucket.RemoveAttachments(context.Background(), "camera/front", nil)
	if err != nil {
		panic(err)
	}
}

File: docs/examples/cpp/src/data_management_attachments.cc
#include <cassert>
#include <reduct/client.h>
#include <set>

using reduct::Error;
using reduct::IBucket;
using reduct::IClient;

int main() {
    // Create a client instance, then get or create a bucket
    auto client = IClient::Build("http://127.0.0.1:8383", {.api_token = "my-token"});
    auto [bucket, create_err] = client->GetOrCreateBucket("my-bucket");
    assert(create_err == Error::kOk);

    // Write attachments for an entry
    IBucket::AttachmentMap attachments = {
        {"schema", R"({"type":"object","version":"1.0"})"},
        {"calibration", R"({"fx":1260.1,"fy":1261.2})"},
    };
    auto err = bucket->WriteAttachments("camera/front", attachments);
    assert(err == Error::kOk);

    // Read attachments
    auto [stored, read_err] = bucket->ReadAttachments("camera/front");
    assert(read_err == Error::kOk);
    assert(stored.contains("schema"));

    // Remove one attachment by key
    err = bucket->RemoveAttachments("camera/front", std::set<std::string>{"calibration"});
    assert(err == Error::kOk);
    auto [stored_after_remove, read_err_after_remove] = bucket->ReadAttachments("camera/front");
    assert(read_err_after_remove == Error::kOk);
    assert(!stored_after_remove.contains("calibration"));

    // Remove all attachments
    err = bucket->RemoveAttachments("camera/front", {});
    assert(err == Error::kOk);
    auto [empty, read_err_empty] = bucket->ReadAttachments("camera/front");
    assert(read_err_empty == Error::kOk);
    assert(empty.empty());

    return 0;
}
